### PR TITLE
Apollocar3d

### PR DESCRIPTION
**Pull request by bertoni9**
Plugin for detecting car keypoints
* conversion from ApolloCar3D format to COCO
* sparsify the original set of keypoints into a set of 24 semantic keypoints
* training and evaluating on the ApolloCar3D dataset

### CODE REVIEW

**Comment by svenkreiss** [comment]
Thanks Lorenzo! I am happy to see this come together.

I am going to change the base branch of this PR to dev.

**Comment by bertoni9** [comment]
Very exciting!
Yes thanks, I meant to merge into the dev myself.

From: Sven Kreiss ***@***.***>
Reply to: vita-epfl/openpifpaf ***@***.***>
Date: Friday, 12 March 2021 at 11:34
To: vita-epfl/openpifpaf ***@***.***>
Cc: Bertoni Lorenzo ***@***.***>, Author ***@***.***>
Subject: Re: [vita-epfl/openpifpaf] Apollocar3d (#366)


Thanks Lorenzo! I am happy to see this come together.

I am going to change the base branch of this PR to dev.

File: openpifpaf/plugins/apollocar3d/.gitignore [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+outputs
+outputs/
+shared/
+decoder.prof
+decoder_flame.svg
+*.onnx
+debug.*.png
+test-clis-*/
+*.mlmodel
+slurm*
```
Comment (svenkreiss) [inline openpifpaf/plugins/apollocar3d/.gitignore:55]: Maybe move this `slurm*` to the main `.gitignore` and remove the file. Multiple `.gitignore` files might be confusing.

File: openpifpaf/plugins/apollocar3d/apollocar3d/apollo_to_coco.py [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        raise Exception('please install matplotlib')
+    bins = np.arange(len(cnt_kps))
+    data = np.array(cnt_kps)
+    plt.figure(1)
+    plt.bar(bins, data)
+    plt.xticks(np.arange(len(cnt_kps), step=5))
+    plt.show()
+
+
+def make_new_directory(dir_out):
```
Comment (svenkreiss) [inline openpifpaf/plugins/apollocar3d/apollocar3d/apollo_to_coco.py:276]: I would never remove the user's directories. Feel free to add timestamps to file names to make them unique or something similar but don't delete anything on the hard drive. Even if that means things get messy.

**Review by svenkreiss — COMMENTED** [review]
Excellent!

Looks great. Just some minor requests below and then this is good to be merged.

**Comment by bertoni9** [comment]
Thank you for the comments, very useful in general!

The function to remove is filter_image or class_aware_sample_weight? For the dataloader, I took the original one and modified the things I needed without changing the rest.

From: Sven Kreiss ***@***.***>
Reply to: vita-epfl/openpifpaf ***@***.***>
Date: Friday, 12 March 2021 at 15:05
To: vita-epfl/openpifpaf ***@***.***>
Cc: Bertoni Lorenzo ***@***.***>, Author ***@***.***>
Subject: Re: [vita-epfl/openpifpaf] Apollocar3d (#366)


@svenkreiss commented on this pull request.

Excellent!

Looks great. Just some minor requests below and then this is good to be merged.

________________________________

In openpifpaf/plugins/apollocar3d/apollocar3d/metrics.py<https://github.com/vita-epfl/openpifpaf/pull/366#discussion_r593186853>:

> +    detections_scaled = []  # detection rate

+    px_ref = 368  # CPM crop size in pixels

+

+    def __init__(self):  # pylint: disable=useless-super-delegation

+        super().__init__()

+

+    def accumulate(self, predictions, image_meta, *, ground_truth=None):  # pylint: disable=arguments-differ

+

+        errors = []

+        detections = []

+        errors_scaled = []

+        detections_scaled = []

+

+        # Filter ground-truth

+        for annotation in ground_truth:  # pylint: disable=too-many-nested-blocks

+            if isinstance(annotation, Annotation):

In for loops in Python, it is often nicer to test for the opposite and do continue. With this one here and two more below, this will reduce your indentations a lot.

________________________________

In openpifpaf/plugins/apollocar3d/apollocar3d/dataloader.py<https://github.com/vita-epfl/openpifpaf/pull/366#discussion_r593187381>:

> +        LOG.info('filter for annotations (min kp=%d) ...', min_kp_anns)

+

+        def filter_image(image_id):

+            ann_ids = self.coco.getAnnIds(imgIds=image_id, catIds=self.category_ids)

+            anns = self.coco.loadAnns(ann_ids)

+            anns = [ann for ann in anns if not ann.get('iscrowd')]

+            if not anns:

+                return False

+            kp_anns = [ann for ann in anns

+                       if 'keypoints' in ann and any(v > 0.0 for v in ann['keypoints'][2::3])]

+            return len(kp_anns) >= min_kp_anns

+

+        self.ids = [image_id for image_id in self.ids if filter_image(image_id)]

+        LOG.info('... done.')

+

+    def class_aware_sample_weights(self, max_multiple=10.0):

You are not using this function, right? Can you remove?

________________________________

In openpifpaf/plugins/apollocar3d/.gitignore<https://github.com/vita-epfl/openpifpaf/pull/366#discussion_r593192897>:

> +*.pl.*

+*.pkl

+*.pkl-*

+*.pkl.*

+*.log

+*.zip

+outputs

+outputs/

+shared/

+decoder.prof

+decoder_flame.svg

+*.onnx

+debug.*.png

+test-clis-*/

+*.mlmodel

+slurm*

Maybe move this slurm* to the main .gitignore and remove the file. Multiple .gitignore files might be confusing.

________________________________

In openpifpaf/plugins/apollocar3d/apollocar3d/apollo_to_coco.py<https://github.com/vita-epfl/openpifpaf/pull/366#discussion_r593195272>:

> +        self.json_file["images"] = []

+        self.json_file["annotations"] = []

+

+

+def histogram(cnt_kps):

+    if plt is None:

+        raise Exception('please install matplotlib')

+    bins = np.arange(len(cnt_kps))

+    data = np.array(cnt_kps)

+    plt.figure(1)

+    plt.bar(bins, data)

+    plt.xticks(np.arange(len(cnt_kps), step=5))

+    plt.show()

+

+

+def make_new_directory(dir_out):

I would never remove the user's directories. Feel free to add timestamps to file names to make them unique or something similar but don't delete anything on the hard drive. Even if that means things get messy.

File: openpifpaf/plugins/apollocar3d/apollocar3d/apollo_to_coco.py (hunk 2 above)
Comment (bertoni9) [reply openpifpaf/plugins/apollocar3d/apollocar3d/apollo_to_coco.py]: You are right. I added an assert for non empty directory, as this is not expected

**Comment by svenkreiss** [comment]
Awesome! This is good. Merging it now.

**Pull request merged**